### ISSUE TITLE
driverless tool gives different result compared to cups

### ISSUE THREAD

**Issue by jschwender** [OpenPrinting/cups-filters#353]
The ppd result is actually different from what is created by cups installing a everywhere driver. As an example for my HP OfficeJet Pro 8100 the resolution is reported 300 150 / 300 300 / 600 600 by cups generated ppd, and 300 300 / 600 600 / 600 600 (draft/normal/high) by driverless ppd. Why? Looks like cups does not use driverless tool like it is mentioned in the man page? Is driverless still needed with most recent cups version?

**Comment by jschwender** [comment]
I have a Canon C3725 here where cups is able to create a ppd, whereas driverless (1.27.4) fails completely: Unable to create ppd file...

**Comment by tillkamppeter** [comment]
There are two different PPD generators, it was due to CUPS not having the PPD generator in its public API, so I copied it for `driverless` and with the time did some changes. So they diverged. Especially I did changes on the resolution handling and they got complicated, which made the answers different. This can also be the cause that a certain bug is only in one of the two.

In CUPS PPD files were deprecated years ago, but only recently a replacement for the classic CUPS drivers consisting of PPDs and filters, Printer Applications, daemons which emulate driverless IPP printers, were introduced. With these CUPS would only need to handle driverless IPP printers and does not need PPDs any more.

As this can lead to the case that the PPD support functionality will get removed from libcups I have copied it over into libppd in cups-filters 2.x and with this cups-filters now contains 2 PPD generators used for the same purpose. I need to join them to one.

To find out about your problem, could you do the following:

Run the command 
```
driverless
```
One of the output lines is the URI of your printer (should contain make/model or the printer's hostname or IP somewhere).

Now run
```
driverless --debug URI > out.txt 2> err.txt
```
with `URI` replaced by the URI from above, and attach the resulting files here.

**Comment by jschwender** [comment]
[err.txt](https://github.com/OpenPrinting/cups-filters/files/6157989/err.txt)

**Comment by tillkamppeter** [comment]
Could you run
```
driverless --debug URI > out.txt 2> err.txt
```
for your Canon C3725 and attach BOTH `out.txt` and `err.txt`? Thanks.

**Comment by jschwender** [comment]
this is the hp printer.
[err.txt](https://github.com/OpenPrinting/cups-filters/files/6336879/err.txt)
[out.txt](https://github.com/OpenPrinting/cups-filters/files/6336884/out.txt)
,well the canon printer wes replaced and is no more here. the replacement is similar, but generating a ppd now at least works with this one:
[err.txt](https://github.com/OpenPrinting/cups-filters/files/6336893/err.txt)
[out.txt](https://github.com/OpenPrinting/cups-filters/files/6336894/out.txt)

**Comment by tillkamppeter** [comment]
If you get at least PPDs for all your printers then probably your only problem now it that you get different choices of resolutions depending on whether you use the PPD file generator of CUPS or of cups-filters. Am I right? Are you able to print in both cases?

I generally plan as part of cups-filters 2.x to merge the two PPD generators. The one in CUPS will go away at a certain point of time but it also exists in the new libppd which contains all PPD file support of CUPS, which is deprecated there. The library is for Printer Applications which retro-fit legacy printer drivers. The two PPD generators, in libcupsfilters and in libppd will get merged to one, which will be part of libppd.

**Comment by jschwender** [comment]
Right, i can print on al the printers, so this is not an urgent matter. I appreciate your plan, makes sense to me. Thank you for your support!

**Comment by tillkamppeter** [comment]
I have made the decision about the PPD generators now, but differently. The PPD generator which will stay is the one in libcupsfilters, the one in libppd I have removed. I did it this way as it uses many auxiliary functions which have nothing to do with PPD files and so are better placed in libcupsfilters and not in libppd and I also do not want libppd depend on libcupsfilters as libcupsfilters already depends on libppd.

The PPD generator in libcupsfilters has some extensions for cups-browsed and it also uses only resolutions which the printer actually annoinces (and not 1/2 of the lowest resolution for draft) as the one of CUPS does.

So I ended up with the one of libcupsfilters.

I also did some clean-up with the human-readable strings.
